### PR TITLE
[DYN-2322] Add PythonNet 3.7 CPython as a selectable engine.

### PR DESCRIPTION
**Pull request by reddyashish**
# Purpose

This PR is a cloned from the original PR(https://github.com/DynamoDS/Dynamo/pull/10472) by @Dewb. 

The auto-completion part for the IronPython is fixed. The IronPythonCompletionData and IronPythonCompletionProvider classes have been moved from the PythonNodeModelsWPF project to the DSIronPython project. We can extend this similarly with the new DSCPython. 

There is only one failing test left that is related to function marshaling: FirstClassFunctions

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo

### CODE REVIEW

File: src/Dynamo.All.sln [hunk 1]
```diff
@@ -85,7 +85,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DynamoCoreWpfTests", "..\te
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "IronPythonTests", "..\test\Libraries\DynamoPythonTests\IronPythonTests.csproj", "{773988FE-EDF6-45CB-A63F-482955EB3553}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DSIronPython", "Libraries\DSIronPython\DSIronPython.csproj", "{9EEF4F42-6B3B-4358-9A8A-C2701539A822}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DSCPython", "Libraries\DSCPython\DSCPython.csproj", "{9EEF4F42-6B3B-4358-9A8A-C2701539A822}"
```
Comment (mjkkirschner) [inline src/Dynamo.All.sln:88]: why was this project renamed instead of a new project being added?

Comment (gregmarr) [reply src/Dynamo.All.sln:88]: @Dewb Looks like the new project was created by copying the file of the old project.  This resulted in two projects with the same GUID, which confuses Visual Studio.  You need to manually change the `<ProjectGuid>` in the new project.

Comment (reddyashish) [reply src/Dynamo.All.sln:88]: Yes, the DSIronPython project entry is also present in line 228. Both the DSIronPython and DSCPython projects have the same GUID. I will change it to a different id.

File: src/Dynamo.All.sln [hunk 2]
```diff
@@ -229,6 +225,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WorkspaceDependencyViewExte
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DocumentationBrowserViewExtension", "DocumentationBrowserViewExtension\DocumentationBrowserViewExtension.csproj", "{BE6D0644-05AF-4580-8597-B95920CE923F}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DSCPython", "Libraries\DSCPython\DSCPython.csproj", "{F1541C2D-80A9-4FE7-8D9E-75A8B9FF3479}"
+EndProject
```
Comment (reddyashish) [inline src/Dynamo.All.sln:229]: @mjkkirschner This end tag was missing between this DSCPython and LibraryViewExtensionMSWebBrowser projects. Now the build is passing.

Comment (reddyashish) [reply src/Dynamo.All.sln:229]: The Eng build is still failing. Will look into it.

File: test/Libraries/DynamoPythonTests/IronPythonTests.csproj [hunk 3]
```diff
@@ -46,18 +46,6 @@
       <HintPath>..\..\..\src\packages\IronPython.2.7.9\lib\net45\IronPython.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="IronPython.Modules, Version=2.7.9.0, Culture=neutral, PublicKeyToken=7f709c5b713576e1, processorArchitecture=MSIL">
```
Comment (reddyashish) [inline test/Libraries/DynamoPythonTests/IronPythonTests.csproj:49]: Removed these IronPython.Modules, IronPython.SQLLite and IronPython.Wpf binaries as they were not being used.

**Comment by mjkkirschner** [comment]
We can’t remove these - patch installer yada yada - if you want we can replace them with empty binaries.

**Comment by mjkkirschner** [comment]
Also I am fairly certain ironpython modules is used to include certain parts of the standard modules - like math - they may not be required if the std library we include includes them but I would be wary about these - anything used by ironpython may only have runtime dependencies - not compile time - so if proving it’s not used will be hard.

**Comment by reddyashish** [comment]
@mjkkirschner Yes, will add them back.

File: src/Libraries/DSIronPython/DSIronPython.csproj [hunk 4]
```diff
@@ -36,6 +36,10 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="ICSharpCode.AvalonEdit">
+      <HintPath>..\..\..\extern\avalonEdit\ICSharpCode.AvalonEdit.dll</HintPath>
```
Comment (aparajit-pratap) [inline src/Libraries/DSIronPython/DSIronPython.csproj:40]: Can we use an AvalonEdit NuGet package instead?

File: src/Libraries/DSIronPython/DSIronPython.csproj [hunk 5]
```diff
@@ -105,9 +115,26 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <EmbeddedResource Include="Resources\namespace.png" />
+  </ItemGroup>
+  <ItemGroup>
+    <EmbeddedResource Include="Resources\privateMethod.png" />
+  </ItemGroup>
+  <ItemGroup>
+    <EmbeddedResource Include="Resources\property.png" />
+  </ItemGroup>
```
Comment (aparajit-pratap) [inline src/Libraries/DSIronPython/DSIronPython.csproj:135]: Any specific reason for moving things from `PythonNodeModelsWpf` to here?

File: src/Libraries/DSIronPython/DSIronPython.csproj (hunk 4 above)
Comment (QilongTang) [reply src/Libraries/DSIronPython/DSIronPython.csproj:40]: What version we are using? Nuget.org shows they only published 4.0 +

File: src/Libraries/DSIronPython/DSIronPython_DynamoCustomization.xml [hunk 6]
```diff
@@ -1,11 +0,0 @@
-<?xml version="1.0"?>
-<doc>
-    <assembly>
-        <name>DSIronPython</name>
-    </assembly>
-    <namespaces>
-        <namespace name="DSIronPython">
-            <category>Core.Scripting</category>
```
Comment (mjkkirschner) [inline src/Libraries/DSIronPython/DSIronPython_DynamoCustomization.xml:8]: @reddyashish why is this being deleted?

File: src/Libraries/DynamoPython/Python.csproj [hunk 7]
```diff
@@ -1,101 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <ImportGroup Label="PropertySheets">
-    <Import Project="$(SolutionDir)Config\CS.props" />
```
Comment (mjkkirschner) [inline src/Libraries/DynamoPython/Python.csproj:4]: so - this project is gone? Why? That means the binary will be gone I guess? A similar problem with the patch installers as mentioned before? Were there public apis here? Is the code that was built here moved to another binary?

File: src/Libraries/DSIronPython/DSIronPython.csproj (hunk 4 above)
Comment (QilongTang) [reply src/Libraries/DSIronPython/DSIronPython.csproj:40]: I just checked, seems we are using file version 4.3.1.9429 which is not an official version from the list https://www.nuget.org/packages/AvalonEdit/. I encourage us move on to use nuget dependencies.

File: src/Libraries/DynamoPython/Python.csproj (hunk 7 above)
Comment (mjkkirschner) [reply src/Libraries/DynamoPython/Python.csproj:4]: oh, was this just like a project holding images?

File: test/Libraries/DynamoPythonTests/CodeCompletionTests.cs [hunk 8]
```diff
@@ -117,7 +117,7 @@ public void CanMatchBasicNumVarSingleLine()
             Assert.IsTrue(matches.ContainsKey("a"));
             Assert.AreEqual("5.0", matches["a"]);
         }
-
+
```
Comment (aparajit-pratap) [inline test/Libraries/DynamoPythonTests/CodeCompletionTests.cs:120]: Can we revert this file if there are no changes to it.

File: src/Libraries/DSIronPython/DSIronPython.csproj (hunk 4 above)
Comment (QilongTang) [reply src/Libraries/DSIronPython/DSIronPython.csproj:40]: I actually have addressed similar issue for fontawesome in my all-you-can-eat-buffet branch. This kind of dech_debt is quite easy. If you guys are OK, I will log a Jira to cover all of these. FYI: @DynamoDS/dynamo

File: src/Libraries/DSCPython/CPythonEvaluator.cs [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        /// <summary>
+        ///     Emitted immediately after execution ends or fails
+        /// </summary>
+        [SupressImportIntoVM]
+        public static event EvaluationEventHandler EvaluationEnd;
+
+        /// <summary>
+        /// Called immediately before evaluation starts
+        /// </summary>
+        /// <param name="engine">The engine used to do the evaluation</param>
```
Comment (aparajit-pratap) [inline src/Libraries/DSCPython/CPythonEvaluator.cs:200]: Remove.

File: src/Libraries/DSCPython/CPythonEvaluator.cs [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+            {
+                EvaluationBegin(EvaluationState.Begin, scope, code, bindingValues);
+            }
+        }
+
+        /// <summary>
+        /// Called when the evaluation has completed successfully or failed
+        /// </summary>
+        /// <param name="isSuccessful">Whether the evaluation succeeded or not</param>
+        /// <param name="engine">The engine used to do the evaluation</param>
```
Comment (aparajit-pratap) [inline src/Libraries/DSCPython/CPythonEvaluator.cs:218]: Remove.

File: src/Libraries/DSCPython/CPythonEvaluator.cs [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+            [ArbitraryDimensionArrayImport] IList bindingValues)
+        {
+            // TODO - it would be nice if users could modify a preference
+            // setting enabling the ability to load additional paths
+
+            // Container for paths that will be imported in the PythonEngine
+            //List<string> paths = new List<string>();
+
+            // Attempt to get the Standard Python Library
+            //string stdLib = pythonStandardLibPath();
```
Comment (aparajit-pratap) [inline src/Libraries/DSCPython/CPythonEvaluator.cs:52]: Do we still need any of these?

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 12]
```diff
@@ -56,13 +72,30 @@ protected override string GetInputTooltip(int index)
...
+            { 
+                Engine = DefaultPythonEngine;
+            }
+
+            Func<string, IList, IList, object> pythonEvaluatorMethod;
+            if (Engine == "CPython3")
+            {
+                pythonEvaluatorMethod = DSCPython.CPythonEvaluator.EvaluatePythonScript;
+            }
+            else if (Engine == "IronPython2")
```
Comment (aparajit-pratap) [inline src/Libraries/PythonNodeModels/PythonNode.cs:86]:
```
else if (Engine == DefaultPythonEngine)
```

File: src/Libraries/PythonNodeModels/PythonNode.cs [hunk 13]
```diff
@@ -56,13 +72,30 @@ protected override string GetInputTooltip(int index)
...
-            Func<string, IList, IList, object> backendMethod =
-                IronPythonEvaluator.EvaluateIronPythonScript;
+            
+            if (String.IsNullOrEmpty(Engine))
+            { 
+                Engine = DefaultPythonEngine;
+            }
+
+            Func<string, IList, IList, object> pythonEvaluatorMethod;
+            if (Engine == "CPython3")
```
Comment (aparajit-pratap) [inline src/Libraries/PythonNodeModels/PythonNode.cs:82]: Can we add a const string for "CPython3" as well.

File: src/Libraries/DSIronPython/DSIronPython.csproj (hunk 4 above)
Comment (QilongTang) [reply src/Libraries/DSIronPython/DSIronPython.csproj:40]: Done. https://jira.autodesk.com/browse/DYN-2639

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 13 above)
Comment (QilongTang) [reply src/Libraries/PythonNodeModels/PythonNode.cs:82]: Please do, I have dependency on this..

File: src/Libraries/DSIronPython/DSIronPython_DynamoCustomization.xml (hunk 6 above)
Comment (reddyashish) [reply src/Libraries/DSIronPython/DSIronPython_DynamoCustomization.xml:8]: Not sure where this xml is used but the python node is being saved in the Script -> Editor folder in the library section(same as before).

File: test/Libraries/DynamoPythonTests/CodeCompletionTests.cs (hunk 8 above)
Comment (reddyashish) [reply test/Libraries/DynamoPythonTests/CodeCompletionTests.cs:120]: Yes removing this file.

File: src/Libraries/PythonNodeModels/PythonNode.cs (hunk 13 above)
Comment (reddyashish) [reply src/Libraries/PythonNodeModels/PythonNode.cs:82]: Added it.

File: src/Libraries/DSIronPython/DSIronPython.csproj (hunk 5 above)
Comment (reddyashish) [reply src/Libraries/DSIronPython/DSIronPython.csproj:135]: These property are being used for the auto-completion feature and we have moved all the auto-completion code to the DSIronPython project. The idea was not to have this PythonNodeModelsWPF project directly depend on any python binaries.

File: src/Libraries/DSIronPython/DSIronPython.csproj (hunk 4 above)
Comment (reddyashish) [reply src/Libraries/DSIronPython/DSIronPython.csproj:40]: Thanks Aaron.

File: src/Libraries/DSCPython/CPythonEvaluator.cs (hunk 11 above)
Comment (QilongTang) [reply src/Libraries/DSCPython/CPythonEvaluator.cs:50]: Same question

**Comment by QilongTang** [comment]
Overall looks good to me and fairly safe to merge into the feature branch for our next moves. 
1. I saw there is a binary removed `DynamoPython`, so we may want to figure out a way to create empty binary for that this year.
2. We can ask BRE to build this branch in the newly setup VMs to see if builds. Seems those VMs already have .NetFramework 4.7.2 installed because it is OOTB with VS2019

@aparajit-pratap @mjkkirschner Any other notes?

File: src/Libraries/DSIronPython/DSIronPython.csproj [hunk 14]
```diff
@@ -36,6 +36,10 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="ICSharpCode.AvalonEdit">
+      <HintPath>..\..\..\extern\avalonEdit\ICSharpCode.AvalonEdit.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
```
Comment (reddyashish) [inline src/Libraries/DSIronPython/DSIronPython.csproj:42]: @mjkkirschner The IronPython SQL, Modules and WPF binaries are already added as references here to the DSIronPython project and the private flag is set to true. So the binaries are copied to the bin folder.

**Comment by reddyashish** [comment]
@QilongTang Yes, I will add an empty project in place of DynamoPython.

**Comment by QilongTang** [comment]
@reddyashish Looks good, can we merge this now?

**Comment by QilongTang** [comment]
@DynamoDS/dynamo I am merging this so that BRE can test branch builds..

File: src/Libraries/DSCPython/CPythonEvaluator.cs (hunk 11 above)
Comment (reddyashish) [reply src/Libraries/DSCPython/CPythonEvaluator.cs:50]: I will address this in the follow-up PR.

Comment (QilongTang) [reply src/Libraries/DSCPython/CPythonEvaluator.cs:50]: Thank you @reddyashish

File: src/Libraries/PythonNodeModelsWpf/PythonNodeModelsWpf.csproj [hunk 15]
```diff
@@ -116,40 +92,21 @@
       <Name>DynamoServices</Name>
       <Private>False</Private>
     </ProjectReference>
-    <ProjectReference Include="..\DSIronPython\DSIronPython.csproj">
-      <Project>{9eef4f42-6b3b-4358-9a8a-c2701539a822}</Project>
-      <Name>DSIronPython</Name>
-      <Private>False</Private>
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/PythonNodeModelsWpf.csproj:122]: @reddyashish FYI

**Pull request merged**